### PR TITLE
Breaking change: Blazor StateManager does not use the provided timeout parameter

### LINKED ISSUE #3800: Breaking change: Blazor StateManager does not use the provided timeout parameter
https://github.com/MarimerLLC/csla/blob/881c262f695cfa02b7e936f123280aadd5fb867f/Source/Csla.Blazor/State/StateManager.cs#L46-L52

As shown the timeout parameter is not used in this method. The method is called by the two `InitializeAsync()` methods where one provides a timeout parameters as well.

I think we have two ways to correct this:
1. Remove the timeout parameter and remove the `InitializeAsync(TimeSpan)` method.
2. We extend `ISessionManager.RetrieveSession()` with an overload which also accepts a timeout and pass it along

I think option 2 would be the best so the user can decide whether he wants to wait or not.
In this regard: Should the `SaveState` also get a timeout parameter with passing along to `SendSession()`?

### PR DESCRIPTION
**Pull request by luizfbicalho**
The previous PR had a merge problem

Closes #3800

### CODE REVIEW

**Comment by luizfbicalho** [comment]
There are 29 places that also have timeout, We could create another issue to certify that all of that places have a cancellation token on interfaces and implementations, and then I could add an extention to my upwork job, what do you think?

**Comment by luizfbicalho** [comment]
@rockfordlhotka  can we finish this one first and then we proceed with the other changes? this way I can estimate all the interfaces and inplementations beforehand and create a new milestone?

**Comment by rockfordlhotka** [comment]
I think that is a wise course of action, yes.

File: Source/Csla.Blazor.WebAssembly.Tests/Csla.Blazor.WebAssembly.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    <ApplicationIcon />
+    <OutputType>Library</OutputType>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="8.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="8.0.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
+    <PackageReference Include="Moq" Version="4.20.70" />
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly.Tests/Csla.Blazor.WebAssembly.Tests.csproj:15]: @rockfordlhotka We/You have to decide if we want to depent on Moq. I personally don't like Moq anymore due to the "sponsor link" incident last year. (c.f. https://github.com/devlooped/moq/issues/1372 and a lot more issues/discussion in th repo)

File: Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      mockServiceProvider.Setup(x => x.GetService(typeof(Csla.Serialization.Mobile.MobileFormatter))).Returns(mockFormatter.Object);
+
+      var mockActivator = new Mock<Csla.Server.IDataPortalActivator>();
+      mockActivator.Setup(x => x.CreateInstance(It.Is<Type>(t => t == typeof(Csla.Serialization.Mobile.MobileFormatter)))).Returns(mockFormatter.Object);
+      mockActivator.Setup(x => x.InitializeInstance(It.IsAny<object>()));
+
+      // Mock IServiceProvider
+      mockServiceProvider.Setup(x => x.GetService(typeof(Csla.Server.IDataPortalActivator))).Returns(mockActivator.Object);
+
+      // Mock IServiceProvider
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs:61]: Dead comment?

File: Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+         // Setup the PROTECTED method to mock
+         .Setup<Task<HttpResponseMessage>>(
+            "SendAsync",
+            ItExpr.IsAny<HttpRequestMessage>(),
+            ItExpr.IsAny<CancellationToken>()
+         )
+         // prepare the expected response of the mocked http call
+         .ReturnsAsync((HttpRequestMessage request, CancellationToken cancellationToken) =>
+         {
+
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs:100]: Empty lines

File: Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    public async Task SendSession_WithCancelledCancellationToken_ShouldThrowTaskCanceledException()
+    {
+      var cts = new CancellationTokenSource();
+      cts.Cancel();
+      await Assert.ThrowsExceptionAsync<TaskCanceledException>(() => _sessionManager.SendSession(cts.Token));
+    }
+
+
+    [TestMethod]
+    public void  RetrieveCAchedSessionSession()
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs:199]: Shouldn't we have two tests here? One checks for `null` return when no session is cached and one which returns a session when it's expected a session is cached?
Without anything this test currently only tests that no exception is thrown in "fresh startup" situation, right?

File: Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+using Csla.Serialization.Mobile;
+using Microsoft.AspNetCore.Components.Authorization;
+
+namespace Csla.Test.State
+{
+  [TestClass]
+  public class SessionManagerTests
+  {
+    private SessionManager _sessionManager;
+    private SessionMessage SessionValue;
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs:25]: Field with `_s` not PascalCase.

File: Source/Csla/State/ISessionManager.cs [hunk 6]
```diff
@@ -18,7 +18,7 @@ public interface ISessionManager
     /// Retrieves the current user's session from
     /// the web server to the wasm client.
     /// </summary>
-    Task<Session> RetrieveSession();
+    Task<Session> RetrieveSession(TimeSpan timeout);
```
Comment (StefanOssendorf) [inline Source/Csla/State/ISessionManager.cs:21]: xml comment missing

File: Source/Csla/State/ISessionManager.cs [hunk 7]
```diff
@@ -27,7 +27,7 @@ public interface ISessionManager
     /// Sends the current user's session from
     /// the wasm client to the web server.
     /// </summary>
-    Task SendSession();
+    Task SendSession(TimeSpan timeout);
```
Comment (StefanOssendorf) [inline Source/Csla/State/ISessionManager.cs:30]: xml comment missing

File: Source/Csla.Blazor/State/StateManager.cs [hunk 8]
```diff
@@ -57,12 +56,12 @@ private async Task GetState(TimeSpan timeout)
     /// at which you know the user is navigating to another
     /// page.
     /// </remarks>
-    public async Task SaveState()
+    public async Task SaveState(TimeSpan timeout)
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor/State/StateManager.cs:59]: To make your life easier with the comments here I suggest using `/// <inheritdoc />` [Docs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#inheritdoc)

File: Source/Csla.Blazor.WebAssembly/State/SessionManager.cs [hunk 9]
```diff
@@ -86,7 +104,30 @@ public async Task<Session> RetrieveSession()
...
+      var cts = new CancellationTokenSource(timeout);
+      return cts.Token;
+    }
+
+    /// <summary>
+    /// Sends the current user's session from
+    /// the wasm client to the web server 
+    /// if SyncContextWithServer is true.
+    /// </summary>
+    public async Task SendSession(CancellationToken ct)
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly/State/SessionManager.cs:130]: Token comment missing

File: Source/Csla.Blazor.WebAssembly/State/SessionManager.cs [hunk 10]
```diff
@@ -44,16 +46,32 @@ public Session GetCachedSession()
     /// the web server to the wasm client
     /// if SyncContextWithServer is true.
     /// </summary>
-    public async Task<Session> RetrieveSession()
+    public async Task<Session> RetrieveSession(TimeSpan timeout)
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly/State/SessionManager.cs:49]: Comment missing. Inhertidoc should work here.

File: Source/Csla.Blazor.WebAssembly/State/SessionManager.cs [hunk 11]
```diff
@@ -44,16 +46,32 @@ public Session GetCachedSession()
...
+        throw new TimeoutException($"{this.GetType().FullName}.{nameof(RetrieveSession)}.", tcex);
+      }
+    }
+
+    /// <summary>
+    /// Retrieves the current user's session from
+    /// the web server to the wasm client
+    /// if SyncContextWithServer is true.
+    /// </summary>
+    public async Task<Session> RetrieveSession(CancellationToken ct)
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly/State/SessionManager.cs:66]: Token comment missing

File: Source/Csla.Blazor.WebAssembly/State/SessionManager.cs [hunk 12]
```diff
@@ -86,7 +104,30 @@ public async Task<Session> RetrieveSession()
     /// the wasm client to the web server 
     /// if SyncContextWithServer is true.
     /// </summary>
-    public async Task SendSession()
+    public async Task SendSession(TimeSpan timeout)
```
Comment (StefanOssendorf) [inline Source/Csla.Blazor.WebAssembly/State/SessionManager.cs:107]: Missing xml doc. Inheritdoc should work here.

**Review by StefanOssendorf — CHANGES_REQUESTED** [review]
Thanks for the re-publish of your PR. I found some minor stuff - docs missing - nothing with the code itself. And an important question for rocky.

**Comment by StefanOssendorf** [comment]
@luizfbicalho Did you send a contributor agreement to rocky? I'll mark the PR as "missing agreement" until I hear from rocky or he removes the tag.

**Comment by luizfbicalho** [comment]
> @luizfbicalho Did you send a contributor agreement to rocky? I'll mark the PR as "missing agreement" until I hear from rocky or he removes the tag.
> 
> Edit: You are a member of the csla developers so I think the agreement is there. 👍

I sent to him, do you want me to send a copy?

File: Source/Csla.Blazor.WebAssembly.Tests/Csla.Blazor.WebAssembly.Tests.csproj (hunk 1 above)
Comment (luizfbicalho) [reply Source/Csla.Blazor.WebAssembly.Tests/Csla.Blazor.WebAssembly.Tests.csproj:15]: @StefanOssendorf  I read about this issue and the latest release https://github.com/devlooped/moq/releases/tag/v4.20.70 removed this with this PR https://github.com/devlooped/moq/pull/1402

Comment (StefanOssendorf) [reply Source/Csla.Blazor.WebAssembly.Tests/Csla.Blazor.WebAssembly.Tests.csproj:15]: That's true. But it's still a loss of trust.

File: Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs (hunk 5 above)
Comment (StefanOssendorf) [reply Source/Csla.Blazor.WebAssembly.Tests/SessionManagerTests.cs:25]: I think you missunderstood me.
`_sessionManager` is correct.
`SessionValue` is not. It should be `_sessionValue`.

**Comment by StefanOssendorf** [comment]
> > @luizfbicalho Did you send a contributor agreement to rocky? I'll mark the PR as "missing agreement" until I hear from rocky or he removes the tag.
> > Edit: You are a member of the csla developers so I think the agreement is there. 👍
> 
> I sent to him, do you want me to send a copy?

No, all good. I found the place where to look for if an agreement was submitted or not :)

File: Source/Csla.Blazor.WebAssembly.Tests/Csla.Blazor.WebAssembly.Tests.csproj (hunk 1 above)
Comment (luizfbicalho) [reply Source/Csla.Blazor.WebAssembly.Tests/Csla.Blazor.WebAssembly.Tests.csproj:15]: What can we do to finalize this?

Comment (luizfbicalho) [reply Source/Csla.Blazor.WebAssembly.Tests/Csla.Blazor.WebAssembly.Tests.csproj]: Changes to NSubstitute

**Review by luizfbicalho — COMMENTED** [review]
all changes made

**Comment by StefanOssendorf** [comment]
Sorry for the delay and thank you for your patience.

LTGM :-)

**Review by StefanOssendorf — APPROVED** [review]

**Pull request merged**